### PR TITLE
ARROW-11718: [Rust] Don't write IPC footers on drop

### PR DESCRIPTION
**Pull request by sfackler**
As discussed in #9520, these destructors shouldn't be there at all.

### CODE REVIEW

File: rust/arrow/src/ipc/writer.rs [hunk 1]
```diff
@@ -434,6 +434,12 @@ impl<W: Write> FileWriter<W> {
 
     /// Write footer and closing tag, then mark the writer as done
     pub fn finish(&mut self) -> Result<()> {
+        if self.finished {
```
Comment (sfackler) [inline rust/arrow/src/ipc/writer.rs:437]: This is just a drive-by fix to defend against double-writes of the footer I noticed while making the change.

Comment (alamb) [reply rust/arrow/src/ipc/writer.rs:437]: this  looks like a good change to to me ( protect people from themselves)

**Review by alamb — COMMENTED** [review]
Thanks for this @sfackler!  Can you remind me again why it is bad to try and write the footer automatically on drop? 

I realize it may not always succeed but on the other hand it also seems to be convenient and not harmful -- people are always free to call `finish` if they wish.

**Comment by sfackler** [comment]
There are two potential sources of error when writing a stream. It could come from the underlying writer (e.g. the TCP socket is closed or the disk is full), in which case the automatic footer write will probably also fail. However, it could come from the code *producing* the data (e.g. it's reading from a database and that connection dies). If that is the case, the Arrow stream is logically incomplete but if a footer is automatically applied on drop, it can appear to a downstream consumer like everything is fine.

**Review by alamb — APPROVED** [review]
@sfackler  that explanation makes sense. Thank you. 

I am concerned that with this change all code that uses the `FileWriter` will need to ensure it calls `finish` before being dropped. 

While the extra call to `finish` might seem redundant at first reading, it is in fact the mechanism the app code uses to signal a successful stream completion. The parquet `FileWriter` has a similar function named `close`: https://github.com/apache/arrow/blob/master/rust/parquet/src/file/writer.rs#L82

@nevi-me  any thoughts / comments? It looks like this behavior came in via https://github.com/apache/arrow/commit/d7e0fa46f32755a9bc1698d79231c0469b7dcb8e

**Comment by alamb** [comment]
I plan to merge this PR on this upcoming weekend barring any additional suggested changes

**Review by nevi-me — APPROVED** [review]
I've checked the reasoning, and I'm happy with it. I am the one who added the destructors, so I'm happy with us removing them.
